File: src/main/java/official/kirizu/kirivers/client/adapter/JdkHttpTransport.java
package official.kirizu.kirivers.client.adapter;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse.BodyHandlers;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

/** Default {@link Transport} using JDK 11+ {@link HttpClient}. Not OkHttp. */
public final class JdkHttpTransport implements Transport {

  private final HttpClient http;
  private final Duration requestTimeout;

  public JdkHttpTransport() {
    this(HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build(), Duration.ofSeconds(60));
  }

  public JdkHttpTransport(HttpClient http, Duration requestTimeout) {
    this.http = Objects.requireNonNull(http, "http");
    this.requestTimeout = requestTimeout == null ? Duration.ofSeconds(60) : requestTimeout;
  }

  @Override
  public HttpResponse execute(HttpRequest request) throws IOException {
    java.net.http.HttpRequest.Builder b =
        java.net.http.HttpRequest.newBuilder(URI.create(request.url())).timeout(requestTimeout);
    for (Map.Entry<String, String> h : request.headers().entrySet()) {
      if (h.getValue() != null) {
        b.header(h.getKey(), h.getValue());
      }
    }
    String method = request.method().toUpperCase(Locale.ROOT);
    byte[] body = request.body();
    if ("GET".equals(method) || "HEAD".equals(method)) {
      b.method(method, BodyPublishers.noBody());
    } else if (body == null || body.length == 0) {
      b.method(method, BodyPublishers.noBody());
    } else {
      b.method(method, BodyPublishers.ofByteArray(body));
    }
    try {
      java.net.http.HttpResponse<byte[]> resp = http.send(b.build(), BodyHandlers.ofByteArray());
      Map<String, List<String>> headers = new LinkedHashMap<>();
      resp.headers().map().forEach(headers::put);
      return new HttpResponse(resp.statusCode(), headers, resp.body());
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new IOException("HTTP request interrupted", e);
    }
  }
}

File: src/main/kotlin/official/kirizu/kirivers/client/JdkHttpTransport.kt
package official.kirizu.kirivers.client

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest as JdkRequest
import java.net.http.HttpResponse as JdkResponse
import java.time.Duration

/**
 * Default [Transport]: JDK [HttpClient]. Follows redirects so private package URLs
 * that bounce to object storage still keep the original query (`exp`/`sig` are already
 * on the request URI).
 */
class JdkHttpTransport(
    private val client: HttpClient = HttpClient.newBuilder()
        .followRedirects(HttpClient.Redirect.NORMAL)
        .connectTimeout(Duration.ofSeconds(30))
        .build(),
    private val requestTimeout: Duration = Duration.ofSeconds(60),
) : Transport {
    override fun execute(request: HttpRequest): HttpResponse {
        val builder = JdkRequest.newBuilder(URI.create(request.url))
            .timeout(requestTimeout)
        request.headers.forEach { (name, value) ->
            if (!RESTRICTED.contains(name.lowercase())) {
                builder.header(name, value)
            }
        }
        val body = request.body
        val publisher = if (body == null) {
            JdkRequest.BodyPublishers.noBody()
        } else {
            JdkRequest.BodyPublishers.ofByteArray(body)
        }
        builder.method(request.method, publisher)
        val jdk = client.send(builder.build(), JdkResponse.BodyHandlers.ofByteArray())
        return HttpResponse(
            status = jdk.statusCode(),
            headers = jdk.headers().map(),
            body = jdk.body() ?: ByteArray(0),
        )
    }

    private companion object {
        val RESTRICTED = setOf("connection", "content-length", "expect", "host", "upgrade")
    }
}
